Add CompositeFormatter class to bokeh>models>widgets>tables.py

jensencl opened the pull request:
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [ ] issues: fixes #9584
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -239,6 +242,29 @@ class BooleanFormatter(CellFormatter):
+        NumberFormatter(format='0[.]0000'),
+        HTMLTemplateFormatter(template='<a href="https:/www.google.com/search?q=<%='
+                                       ' manufacturer %>+<%= model %>" target="_blank">'
+                                       '<%= value %></a>'),
```
Would ask for some formatting changes here, also there is a sphinx directive that will get this formatted in the docs:
```
Example:

    .. code-block:: python
 
    example_formatter = CompositeFormatter(formatters=[
        NumberFormatter(format='0[.]0000'),
        HTMLTemplateFormatter(template='<a href="https:/www.google.com/search?q=<%='
                                       ' manufacturer %>+<%= model %>" target="_blank">'
                                       '<%= value %></a>'),
    ])
```
Also, I think it is important to specify the order in which the formatters will be applied. Is the first formatter closest to the input (applied first) or closest to the output (applied last).

jensencl replied:
I think I am following what you are suggesting here. To do this, would you recommend using something like an ordered list or set to state the order in which the formatters will be applied in the script, or would it be better to write something into the tables.py that would allow the user to set the declare the order they want to prioritize their formatters? Or am I way off-base

I was actually thinking about something similar to this quite a bit last night (I had one of those things are rattling around in my head nights and couldn't sleep). I am probably hugely overthinking this and unintentionally making it more complicated than it is and I think explicitly stating the order would negate this thought, but here is what I had rolling around in the empty space between my ears....

There are combinations of formatters that don't make sense (at least to me) to ever be used together. The most obvious combination that comes to mind for me is If you're using a ScientificFormatter, you wouldn't be using a NumberFormatter or BooleanFormatter. If a user were to try to enter these combinations, is this something that I should consider including as a returning an error for, or are we best off at this point getting a working CompositeFormatter out there, then think about raising errors if necessary?

bryevdv replied:
@jensencl The formatters property will preserve an ordered list/sequence already. My comment about stating the order was really about adding a sentence in the docstring that says which one of the formatters in that ordered sequence *we choose* to apply. Some people (esp mathematicians e.g) might expect that the last formatter in the list is the first formatter that the values is passed to. We just need to pick a convention, and document what we chose. 

Regarding the possible incompatible types. That's definitely possibility. Roughly, I think the "input type" for all but the first formatter needs to be "string". There are ways we could attempt to detect and warn about this situation. But I would say we can defer a decision about what, if anything, to do about that. The upside of a modular "building block" architecture like Bokeh's models is that it is very flexible to put all the pieces together in different ways. The downside is that not all combinations of building blocks make sense. It's a judgment call to decide what we can/want to try and detect and warn about.

jensencl replied:
Cleaned up the formatting and added a short sentence to address the order of implementation.

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -14,6 +14,8 @@
 import logging # isort:skip
 log = logging.getLogger(__name__)
 
+from Lib.typing import Sequence
```
In this case `Sequence` is not from Python, it is one of the Bokeh  property types, so it should go in the list of all the property imports below.

jensencl replied:
Under `core.properties` on line 229 I am finding is `Seq` listed under `__all__ = (` , I didn't see one in the `core.properties.py` named `Sequence`. I am probably just overlooking it though? `Seq` looks like it is a `ContainerProperty`.

bryevdv replied:
@jensencl Apologies, `Seq` is actually what you want, I misspoke. That Bokeh property type will accept most any sequence type (lists, tuples, arrays, series, etc.)

bryevdv commented:
Thanks for starting the PR @jensencl ! I have a couple of prelim comments on the Python pat so far then we can dive in to adding the JS part.

jensencl commented:
@bryevdv I'm sorry I am so slow on this. I am doing my best to try to adhere to the project conventions. This is the first time I have really dealt with trying to think through all the operations being performed while still adhering to a guided project documentation format. 

Thank you very much for being patient and helping me with this. I am actually learning quite a bit how everything works behind the scenes as I try to address everything.

bryevdv commented:
@jensencl There's really no hurry or set timeframe, please work at whatever pace works for you!

jensencl commented:
@bryevdv I went ahead and pushed up the tables.py with the updated class. Hopefully it looks a little better now.

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -1,740 +1,764 @@
+import logging # isort:skip
+log = logging.getLogger(__name__)
+
+# from typing import Sequence
```
This line should be deleted entirely

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -1,740 +1,764 @@
+    Int,
+    List,
+    Override,
+    String,
```
`Seq` needs to be added to this list of imports

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -1,740 +1,764 @@
+    .. code-block:: python
+
+        example_formatter = CompositeFormatter(formatters=[StringFormatter(font_style="bold"), HTMLTemplateFormatter(
+        template='<code><%= value %></code>')])
```
nit: It's common to indent a line continuation, and we definitely adhere to that style throughout. To be consistent I would suggest something like:
```
        example_formatter = CompositeFormatter(formatters=[
            StringFormatter(font_style="bold"), 
            HTMLTemplateFormatter(template='<code><%= value %></code>')
        ])
```

bryevdv commented on bokeh/models/widgets/tables.py:
```diff
@@ -1,740 +1,764 @@
+        formatting capabilities (BooleanFormatter, CellFormatter, DateFormatter,
+        HTMLTemplateFormatter, NumberFormatter, ScientificFormatter, StringFormatter,
+        to a single TableColumn. Formatters are implemented in the order they are 
+        listed, i.e: from "(" to ")".
```
There may not be `(` e.g. if the sequence is a list, or an array of objects, so that might be confusing. Suggest:
> Formatter are applied in the order they appear in the sequence.

bryevdv commented:
@jensencl Thanks for the update! I've left a few more small comments (the missing `Seq` import is causing all the test failures)

jensencl commented:
@bryevdv All fixed and updated in the PR file. I tried running it again using the data_tables.py file in the supplied examples with the dev package and I am still getting the same thing. I am outputting a "data_tables.html" file, but the file remains blank. If I run it without the "example_formatter" everything comes out fine. Is it something wrong with my code, or have I forgotten something?

I added the file "TestFileOutput.html" https://github.com/jensencl/bokeh/blob/master/TestFileOutput.html to the main folder of my PR repo. If there is something going on here that I am overlooking, I am sure you'll catch it much faster than I would. It looks like it is assigning the id: 1004 to the CompositeFormatter and applying that to the "Displacement" column as I have written it, but something is leaving me with a blank page in the html as shown in the image below. At the very bottom is the file I have output without using the CompositeFormatter. You can see that one outputs the file as expected. Is this possibly just becuase I have not updated the bokeh.js yet?


![data_table html-Capture](https://user-images.githubusercontent.com/60002857/73170448-68901300-40cc-11ea-85e5-6d40f9db280d.JPG)
![data_table html(No CompFormatter-Capture](https://user-images.githubusercontent.com/60002857/73170602-da685c80-40cc-11ea-9025-cf32ea953830.JPG)

bryevdv commented:
> Is it something wrong with my code, or have I forgotten something?

@jensencl Nothing wrong with your code, but the Python changes are only half the story. Without the corresponding BokehJS implementation, there is nothing that actually does the work described by the new Python class. I figure we'd go though and get the Python side all done and then focus on the BokehJS bits (which should be quite small)

That said, something *has* gone wrong with the Git aspect of the PR. I'm not sure what caused it without seeing your git commend history, but there is now a branch conflict and the diff looks a little odd.  (Did you do a "rebase" by chance?) In any event, I would say rather than try to resolve the conflict, it's probably simpler just to close this PR and open a new one, but from a branch on your repo this time. That's actually a good idea anyway, it seems that the GH badge that report CI build status will report on *any* master branch status, not just the main repo. So currently our badge is reporting "failing" even though it is not, really. Does that sound like a plan? Basically the steps are:

* git checkout -b new_branch_name
* make changes on the branch
* push the branch up to GitHub
* make PR from the new branch (instead of from your master branch as this PR is)

That's a very high-level, rushed overview. Let  me know if I can help with any specific steps, or find some additional resources!

jensencl commented:
@bryevdv When I try to push up to GitHub I am getting an error related to django.conf. I have toyed with it for about an hour, but haven't been able to figure it out. Can you possibly point me in the right direction?

07:13:45.831: [bokeh] git -c core.quotepath=false -c log.showSignature=false add --ignore-errors -A -f -- .idea/inspectionProfiles/profiles_settings.xml .idea/modules.xml .idea/.gitignore .idea/vcs.xml .gitignore .idea/misc.xml .idea/bokeh.iml
07:13:46.083: [bokeh] git -c core.quotepath=false -c log.showSignature=false commit -F C:\Users\cj3\AppData\Local\Temp\git-commit-msg-.txt --
Traceback (most recent call last):
  File "c:\users\cj3\appdata\roaming\python\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\cj3\AppData\Roaming\Python\Python37\Scripts\py.test.exe\__main__.py", line 7, in <module>
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\config\__init__.py", line 73, in main
    config = _prepareconfig(args, plugins)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\config\__init__.py", line 224, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\hooks.py", line 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\manager.py", line 93, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\manager.py", line 87, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\callers.py", line 203, in _multicall
    gen.send(outcome)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\helpconfig.py", line 89, in pytest_cmdline_parse
    config = outcome.get_result()
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\config\__init__.py", line 794, in pytest_cmdline_parse
    self.parse(args)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\config\__init__.py", line 1000, in parse
    self._preparse(args, addopts=addopts)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\config\__init__.py", line 948, in _preparse
    self.pluginmanager.load_setuptools_entrypoints("pytest11")
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\pluggy\manager.py", line 299, in load_setuptools_entrypoints
    plugin = ep.load()
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\importlib_metadata\__init__.py", line 94, in load
    module = import_module(match.group('module'))
  File "c:\users\cj3\appdata\roaming\python\python37\lib\importlib\__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\_pytest\assertion\rewrite.py", line 143, in exec_module
    exec(co, module.__dict__)
  File "c:\users\cj3\appdata\roaming\python\python37\lib\site-packages\django_webtest\__init__.py", line 4, in <module>
    from django.conf import settings
ModuleNotFoundError: No module named 'django.conf

jensencl commented:
I did some testing and I believe that the `ModuleNotFoundError: No module named 'django.conf` I am getting is causing the failed checks. I ignored the Git hooks for a commit and did not get this error. I must not have had something set up correctly before with the hooks. I have dug around in this for an hour or two and am rather perplexed at this point. I even tried reinstalling everything from square-one.

bryevdv commented:
@jensencl can you try `pip install channels` into your environment? That's a recent test dependency that is not conda-installable yet. I woudn't have thought it would much with the pre-commit hook codebase tests, but perhaps.

jensencl commented:
still the same thing. I know I have to be doing something wrong here, but I just can't for the life of me figure it out. Just to be sure I have the hooks set up right, there should be 'pre-commit' and 'post-commit' in the directory "....\bokeh\.git\hooks", correct?

bryevdv commented:
@jensencl should be, at this point I would suggest just deleting/removing the git hooks and seeing if that unblocks things?

jensencl commented:
@bryevdv Looks like we are in business. Says there are no conflicts. New pull request is here: https://github.com/bokeh/bokeh/pull/9619

bryevdv commented:
superseded by #9619
